### PR TITLE
xco2 derivation

### LINKED ISSUE #912: XCO2 derivation
Now that XCH4 is merged, let's add XCO2 in the same fashion.

Required for https://github.com/ESMValGroup/ESMValTool/pull/1914 .
Core Pull Request: #913

### PR DESCRIPTION
**Pull request by bettina-gier**
Before you start, please read our [contribution guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html).

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #912

Now that XCH4 is merged, let's add XCO2!

Sample recipe with some data available on mistral:
```
xco2_test:
    description: XCO2 test
    variables:
      xco2:
        project: CMIP6
        mip: Amon
        exp: esm-hist
        start_year: 2003
        end_year: 2005
        derive: true
    additional_datasets:
      - {dataset: CDS-XCO2, project: OBS, type: sat, version: L3,
         tier: 3, derive: false}
      - {dataset: CNRM-ESM2-1, ensemble: r1i1p1f3, grid: gr}
      - {dataset: GFDL-ESM4, ensemble: r1i1p1f1, grid: gr1}
      - {dataset: MPI-ESM1-2-LR, ensemble: r1i1p1f1, grid: gn}
    scripts: null
```

Using some additional fixes, and other models, as well as diagnostics from an open pull request for a comparison:
xco2_esm-hist_global_time_series_panels_2003-2014_obs

### CODE REVIEW

**Review by jvegasbsc — APPROVED** [review]
Code-wise, it looks great. Thanks!

**Review by valeriupredoi — APPROVED** [review]
nice plots&code @bettina-gier - could you pls ping @hb326 or @axel-lauer or whoever else you reckon as appropriate for a sci review? :beer:

**Comment by bettina-gier** [comment]
Both of them work, I just don't know if they're back from christmas vacation - but consider them pinged with your message ;)
I did add Birgit as a reviewer when I opened the pull request.

File: esmvalcore/preprocessor/_derive/xco2.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    def calculate(cubes):
+        """Calculate the column-averaged atmospheric CO2 [1e-6]."""
+        co2_cube = cubes.extract_strict(
+            Constraint(name='mole_fraction_of_carbon_dioxide_in_air'))
+        print(co2_cube)
+        hus_cube = cubes.extract_strict(Constraint(name='specific_humidity'))
+        zg_cube = cubes.extract_strict(Constraint(name='geopotential_height'))
+        ps_cube = cubes.extract_strict(Constraint(name='surface_air_pressure'))
+
+        # Column-averaged CH4
```
Comment (hb326) [inline esmvalcore/preprocessor/_derive/xco2.py:33]: That should read here "Column-averaged CO2", right?

File: esmvalcore/preprocessor/_derive/xco2.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+"""Derivation of variable ``xch4``."""
+
```
Comment (hb326) [inline esmvalcore/preprocessor/_derive/xco2.py:2]: Should read here "Derivation of variable ''xco2''.", right?

File: tests/unit/preprocessor/_derive/test_xco2.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+                       units='m')
+    ps_cube = get_cube([[[100000.0]]], air_pressure_coord=False,
+                       standard_name='surface_air_pressure', var_name='ps',
+                       units='Pa')
+    return iris.cube.CubeList([co2_cube, hus_cube, zg_cube, ps_cube])
+
+
+def test_xch4_calculate(cubes):
+    """Test function ``calculate``."""
```
Comment (hb326) [inline tests/unit/preprocessor/_derive/test_xco2.py:31]: Should this here be called "test_xco2_calculate(cubes):"?
Or are you just adding the unit test for xco2 to the xch4 unit test?

**Comment by hb326** [comment]
@bettina-gier: I found some "ch4" reference remnants in the code, of which I am not sure if they should still be in there or if they are just leftovers from the xch4 derivation code that you used as a guideline for the xco2 code.
I also ran the test recipe to derive xco2 from the models. It seemed to work fine, although one of the models you had in the list does not have data on DKRZ anymore...

**Comment by bettina-gier** [comment]
You're right they're leftovers, thanks for spotting these! Maybe I should have just used find and replace =D
and yes I have noticed the one model having missing data as well.. I will write that mail soon, we realized in a group meeting today that most of us had the same issue so I'll included samples of missing data from all of us working on different topics.

**Review by hb326 — APPROVED** [review]

**Comment by hb326** [comment]
Not sure why there is now a check failing...

**Comment by valeriupredoi** [comment]
it's the hardcoded 2020 crap year in era5 that has been fixed upstream :+1:

**Pull request merged**